---
title: "ClustersWithGO"
author: "Cas Baptist"
date: "4/6/2022"
output: html_document
---

***

##### The following code associates annotations from Gene Ontology to different clusters from the yeast genome using data from DEE2  

***

</br>


### Prerequisite libraries
```{r Libraries, echo=TRUE}
suppressPackageStartupMessages({c(library(R.utils), library(dplyr), library(tidyr), library(data.table), library(RColorBrewer), library(gplots), library(Polychrome), library(tidyverse), library(edgeR), library(GO.db), library(gage), library(gProfileR), library(factoextra), library(NbClust), library(org.Sc.sgd.db),library(stringi))})

```
</br>

###  1. Download and unzip the data (ok)
```{r Download Data, echo=TRUE}
# put in an if statement to check if object exists

if (!file.exists("Data/hsapiens_rand_se.tsv.bz2")){
    download.file("http://ziemann-lab.net/public/cas/rand/hsapiens_rand_se.tsv.bz2", 
              destfile = "Data/hsapiens_rand_se.tsv.bz2")}

# fread error msg: This file is very unusual: it ends abruptly without a final newline, and also its size is a multiple of 4096 bytes. Please properly end the last row with a newline using for example 'echo >> file' to avoid this  time to copy. 

# needed to go back to bunzip2
if (!file.exists("Data/hsapiens_rand_se.tsv")){
    bunzip2("Data/hsapiens_rand_se.tsv.bz2", destname="Data/hsapiens_rand_se.tsv")}

orgData <- fread("Data/hsapiens_rand_se.tsv", sep = '\t', header = FALSE)

head(orgData)
```
</br>

### 2. Download, read, and unzip the QC data (ok)
```{r QC Data, echo=TRUE}
if (!file.exists("hsapiens_rand_qc.tsv.bz2")){
    download.file("http://ziemann-lab.net/public/cas/rand/hsapiens_rand_qc.tsv.bz2", 
              destfile = "hsapiens_rand_qc.tsv.bz2")}

qcData <- fread("hsapiens_rand_qc.tsv.bz2", sep = '\t', 
  col.names = c("Database", "Description", "Details"))

head(qcData)
```
</br>

### 3. Filter and quantify databases according to QC_SUMMARY = PASS. (ok)
```{r QC Data Summary, echo=TRUE}

# count the total number of QC_SUMMARY entries for checking
qcSummary <- qcData[qcData$Description == 'QC_SUMMARY',]
totalSummary <- nrow(qcSummary)

# count the total number of databases marked as PASS
qcPass <- filter(qcData, grepl("PASS", Details))
totalPass <- nrow(qcPass)

```
</br>

### 4. Select all the databases from org_se.tsv with QC_SUMMARY == 'PASS' (ok)
```{r Filter yeast bulk data, echo=TRUE}

# filter all the sample names with QC_SUMMARY == 'PASS' from step 3
databasePass <- filter(qcData, grepl("PASS", Details)) 

# convert the rows into string and store on a list to use for filtering Yeast data
databasePassList <- as.list(as.character(databasePass$Database))

# Filter Yeast data using generated list
orgPass <- orgData[orgData$V1 %in% databasePassList,] 

# Convert format from long to wide
orgPassWide <- orgPass %>% pivot_wider(names_from = "V1", values_from = "V3")

# Convert tibble to data frame and assign column 1 as rowname
orgPassWide <- as.data.frame(orgPassWide)
rownames(orgPassWide) <- orgPassWide[,1]
orgPassWide <- orgPassWide[,-1]

orgPassWide[1:10,1:6]

```
</br>

### 5. Aggregate the SRR run data to SRX experiment (ok)
```{r SRR to SRX Aggregation, echo=TRUE}

# Download and read Metadata summary

if (!file.exists("hsapiens_metadata.tsv.cut")){
    download.file("http://ziemann-lab.net/public/cas/rand/hsapiens_metadata.tsv.cut", 
              destfile = "hsapiens_metadata.tsv.cut")}
    
orgMetadata <- read.csv("hsapiens_metadata.tsv.cut", sep = '\t')

# Filter metadata to include only "passed" samples
orgMetadataPass <- orgMetadata[which(orgMetadata$SRR_accession %in% databasePassList),]

# Assign column 1 as rownames 
rownames(orgMetadataPass) <- orgMetadataPass[,1]
orgMetadataPass <- orgMetadataPass[,-1]

# Sequencing Run aggregate function
srxAgg <- function(x,counts="GeneCounts") {
    IDX=which(names(x) %in% "GeneCounts")
    mds<-x$MetadataSummary
    n=nrow(x[[IDX]])
    SRX_dat <- vapply(X=unique(mds$SRX_accession) ,function(srx) {
        srrs<-rownames(mds)[which(mds$SRX_accession %in% srx)]
        if (length(srrs)>1) {
            rowSums(x[[IDX]][,srrs])
        } else {
            x[[IDX]][,srrs]
        }
    } , numeric(n))
    rownames(SRX_dat) <- rownames(x[[IDX]])
    colnames(SRX_dat) <- unique(mds$SRX_accession)
    SRX_dat
}

# Put the filtered org GeneCount data and the filtered metadata dataframes into a list
orgCountMetadata <- list(GeneCounts = orgPassWide, MetadataSummary = orgMetadataPass)

# Apply both dataframes to the function
agg <- srxAgg(orgCountMetadata)

# Check for samples with zero total counts
colSums(agg) == 0

# Histogram showing samples (columns) counts
hist(colSums(agg), breaks = 30)

# Filtering samples with more than 1 million reads
aggFilterCols <- agg[, which(colSums(agg) >= 1e6)]

# Filtering genes with more than zero counts 
aggFilterRows <- agg[which(rowSums(agg) > 0), ]

# No difference between the filtered and unfiltered columns  
dim(aggFilterCols)
dim(agg)

# Rows ommited for genes with more than zero counts
dim(aggFilterRows)

# Remove unneeded data
remove(orgMetadata)

```
</br>

### 6. Normalisation of Data (ok)

```{r eval=FALSE, include=FALSE}
# NORMALISATION FOR LIBRARY SIZE BIAS (cpm)
cpm <- cpm(agg)
# add a small number to the entire dataframe to account for zero reads
cpm <- cpm + 0.01
# Filtering to remove lowly expressed genes
thresh <- cpm > 0.6
# check if there are zero values. Result should be FALSE.
all(apply(apply(cpm, 2, function(x) x==0), 2, any))

# Keep genes that have at least 2 TRUEs in each row of thresh
keepthresh <- rowSums(thresh) >= 2

# Subset the rows of data to keep the more highly expressed genes
keepCount <- cpm[keepthresh,]
summary(keepthresh)
dim(keepCount)

# A CPM of 0.6 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.6. You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.

# Note: When in doubt, a threshold of 1 CPM in at least minimum group sample size is a good rule of thumb.

# Check whether our threshold of 0.6 does indeed correspond to a count of about 10-15
# Add a vertical line at 0.6 CPM, and a horizontal line at Counts = 10
plot(cpm[,1], agg[,1], ylim=c(0,50), xlim=c(0,3), abline(v=0.6, h=10, col="blue"), ylab = "Counts", xlab = "CPM", main = "Counts-per-million (CPM) vs Counts")

# Convert counts to DGEList object (from edgeR library)
dgeObj <- DGEList(keepCount)

#NORMALISATION FOR COMPOSITION BIAS (TMM normalisation)
# Apply normalisation to DGEList object
dgeObj <- calcNormFactors(dgeObj) 
# dgeObj$samples

# Get log2 counts per million
normAgg <- cpm(dgeObj,log=TRUE)

# Check distributions of samples
hist(normAgg)
hist(keepCount)

all(apply(apply(normAgg, 2, function(x) x==0), 2, any))

```
</br>


### 8. Heirarchical Clustering with PASS only databases (ok)
```{r echo=TRUE}

# Applying Hierarchical Clustering after normalization
distClust <-as.dist(1-cor(t(normAgg), method="spearman"))
hClust <- hclust(distClust , method="complete")

# optimizing the cluster size
clusterPass <- cutree(hClust, h=max(hr$height/1.64))
clusterPasslength <- length(unique(cluster))

if (!exists("P100")){
  P100 <- createPalette(100,  c("#ff0000", "#00ff00", "#0000ff"))}

clusterCols <- P100[1:clusterPasslength]
myClusterSideBar <- clusterCols[clusterPass]
colfunc <- colorRampPalette(c("blue", "white", "red"))
if (!file.exists("Data/org_se.tsv")){
  write.table(cluster,file="Data/org_pass.txt",quote=F,sep="\t")}

# create a 5 x 5 inch png image
png("Data/org_heatmaps_PASS_normalization3.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)          

# create the heatmap
heatmap.2(normAgg, main="org PASS Samples",  Rowv=as.dendrogram(hr),
          dendrogram="both", scale="column", col = colfunc(25), trace="none", 
          RowSideColors= myClusterSideBar, margins = c(5,5))

dev.off()
```
</br>



### 9. Gene Ontology and Gene Data (ok)
```{r}
# Get the GO Annotation File (GAF) format 2.1 of org
if (!file.exists("Homo_sapiens.GRCh38.ncrna.fa.gz")){
    download.file("http://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz", 
              destfile = "Homo_sapiens.GRCh38.ncrna.fa.gz")}
if (!file.exists("Homo_sapiens.GRCh38.ncrna.fa.gz")){
    gunzip("Homo_sapiens.GRCh38.ncrna.fa.gz", destname="omo_sapiens.GRCh38.ncrna.fa.gz")}

orgGO <- fread("Homo_sapiens.GRCh38.ncrna.fa.gz", sep = '\t', header = FALSE)

# Subset unique GeneIDs (V11) and GO IDs (V5) in a dataframe. Select the first gene synonym from column V11 and delete the rest. Use this as the GeneID column. Add another column with 1 in each row indicating the relationship of the Gene ID to the GO ID

orgGOmatrix <- data.frame(unique(orgGO[,c(11,5)]))
orgGOmatrix$GeneID <- gsub("\\|.*", "", orgGOmatrix$V11)
orgGOmatrix$V3 <- as.numeric(paste(1, orgGOmatrix$V3))
orgGOmatrix <- orgGOmatrix[,-1]

# Convert the dataframe to wide format with Gene ID as column 1 and GO ID as column names 
orgGOmatrixwide <- orgGOmatrix %>% pivot_wider(names_from = "V5", values_from = "V3")
orgGOmatrixwide <- as.data.frame(orgGOmatrixwide)
colnames(orgGOmatrixwide)[1] <- "GeneID"

# convert NA to Zero
orgGOmatrixwide[is.na(orgGOmatrixwide)] <- 0

# save the dataframe to an RDS file
saveRDS(orgGOmatrixwide, "orgGOmatrixwide.rds")
head(orgGOmatrixwide)

# Remove unneeded data
remove(orgGOmatrix)

```
</br>


### 10. Blinding genes using random strings
```{r}
#Fraction of the total number of genes to be blinded
fracBlind <- 0.10
numGenes <- round(nrow(agg)*fracBlind)

blind <- sample(rownames(agg), numGenes)

rand <- stri_rand_strings(numGenes, 15, pattern = "[A-Za-z]")
rand <- paste("blind_", rand, sep = "")
dictBlind <- data.frame(blind, rand)

# save the dictionary into an RDS file
saveRDS(dictBlind, "dictBlind.rds")

aggRand <- agg

indBlind <- which(rownames(aggRand) %in% blind)

rownames(aggRand)[indBlind] <- rand

```
</br>


### 11. Functions for Cluster Analysis 
```{r echo=TRUE}

# This function's output is a nested list with of the genes grouped per cluster and their corresponding correlation values. Inputs:
# x = normAgg (normalized gene counts from RNA seq)
# y = clusters (matrix of genes w/ cluster number)
# clust_total = total number of clusters

corrClust <- function(x, y, clust_total){
  corrClust <- list()
  for (i in 1:clust_total){
    clusterX <- y[y$ClusterNumber == i,]
    cluster_list <- as.list(clusterX$GeneID)
    cluster <- x[rownames(x) %in% cluster_list,]
    corr_result <- cor(t(cluster))
    corrClust[[paste0("Cluster", i)]] <- corr_result
  }
  return(corrClust)
}

corrAllClusters <- corrClust(normAgg, clusters, 100)


# This function's output is a list of data frames containing all GO terms associated with the genes belonging to a cluster. Input:
# x = orgGOmatrixwide (matrix of genes belonging to a GO term)
# y = clusters (matrix of genes w/ cluster number)
# clust_total = total number of clusters

GOperClust <- function(x,y,clust_total){
  GOClust <- list()
  for (i in 1:clust_total){
    clusterX <- y[y$ClusterNumber == i,]
    cluster_list <- as.list(clusterX$GeneID)
    cluster_GOterms <- x[x$GeneID %in% cluster_list,]
    rownames(cluster_GOterms)<- cluster_GOterms[,1] 
    cluster_GOterms[,1] <- c()
    cluster_GOterms <- cluster_GOterms[,which(colSums(cluster_GOterms) > 0)]
    GO_cl[[paste0("Cluster", i)]] <- cluster_GOterms
  }
  return(GOClust)
}

GOallClusters <- GOperClust(orgGOmatrixwide, clusters, 100)


# This function's output is a nested list of genes belonging to a cluster with their corresponding weighted correlations. Input:
# gene_list = list of genes to be correlated; corrClust = correlation values for cluster
# cl_GO = GO terms for cluster

weightcorrClust <- function(gene_list, corrClust, cl_GO){
  weightcorrClust <- list()
  
  for (i in gene_list){
    corr_value <- corrClust[i,]
    corr_value_df <- as.data.frame(corr_value)
    weighted_go <- merge(corr_value_df, cl_GO, by = "row.names")
    rownames(weighted_go) <- weighted_go[,1]
    weighted_go[,1] <- c()
    weighted_go <- weighted_go[,1]*weighted_go[,2:ncol(weighted_go)]
    normalized_weighted_go <- colSums(weighted_go)/nrow(weighted_go)
    weightcorrClust[[i]] <- normalized_weighted_go
  }
  setNames(weightcorrClust, paste0(i))
  return(weightcorrClust)
}

cluster1GO <- weightcorrClust(corrblind1, corrClust1, cluster1GO)


# Sum of corr values are used as a denominator in normalized_weighted_go
weightcorr2Clust <- function(gene_list, corrClust, cl_GO){
  weightcorrClust <- list()
  
  for (i in gene_list){
    corr_value <- corrClust[i,]
    corr_value_df <- as.data.frame(corr_value)
    weighted_go <- merge(corr_value_df, cl_GO, by = "row.names")
    rownames(weighted_go) <- weighted_go[,1]
    weighted_go[,1] <- c()
    # this is where this differs from the original function. The sum of the correlation values for         gene1 in the gene_list is used as the denominator for the normalized_weighted_go object instead of     the nrow of the dataframe. This follows the formula of the weighted average, assuming that the corr     values are the "weights" 
    corr_colsum <- sum(weighted_go[,1])
    weighted_go <- weighted_go[,1]*weighted_go[,2:ncol(weighted_go)]
    normalized_weighted_go <- colSums(weighted_go)/corr_colsum
    weightcorrClust[[i]] <- normalized_weighted_go
  }
  setNames(weightcorrClust, paste0(i))
  return(weightcorrClust)
}

cluster1GO2 <- weightcorr2Clust(corrblind1, corrClustuster1, cluster1GO)




```
</br>

### 12. Cluster Analysis
```{r}
# Prepare Cluster data frame
cluster <- as.data.frame(cluster)
colnames(cluster) <- "ClusterNumber"

# make a new column, GeneID, from rownames
cluster$GeneID <- rownames(clusters)

# Grouping genes by cluster (using cluster 1)
cluster1 <- cluster[cluster$ClusterNumber == 1,] 
cluster1_list <- cluster1$GeneID

# look for cluster gene match in GO dictionary and subset nonzero columns 
cluster1GO <- orgGOmatrixwide[orgGOmatrixwide$GeneID %in% cluster1_list,] 
cluster1GO_nzero <- cluster1GO[, !apply(cluster1GO == 0, 2, all)]

# Correlation in Cluster 1  (Function counterpart: corrClust)
cluster1 <- normAgg[rownames(normAgg) %in% cluster1_list,]
corrCluster1 <- cor(t(cluster1))

# GO terms for cluster 1 (Function counterpart: GOperClust)
cluster1GO <- orgGOmatrixwide[orgGOmatrixwide$GeneID %in% cluster1_list,]
rownames(cluster1GO)<- cluster1GO[,1] 
cluster1GO[,1] <- c()
cluster1GO <- cluster1GO[,which(colSums(cluster1GO) > 0)]

# weighted correlation for Cluster 1 (Function counterpart: weightcorrCluster)
corrblind1 <- cluster1_list[cluster1_list %in% dictBlind$blind]
geneblind1 <- corrblind1[1]

# Subset the row which contains geneblind1 and turn it into a data frame
corr_value_blind1 <- corrCluster1[geneblind1,]
corrblind1df <- as.data.frame(corr_value_blind1)

# Merge the GO terms data frame for cluster 1 and the correlation values for geneblind1 
weightGOClust1 <- merge(corrblind1df, cluster1GO, by = "row.names")
rownames(weightGOClust1) <- weightGOClust1[,1]
weightGOClust1[,1] <- c()
corr_colsum <- sum(weightGOClust1[,1])
weightGOClust1 <- weightGOClust1[,1]*weightGOClust1[,2:ncol(weightGOClust1)]
normweightGOClust1 <- colSums(weightGOClust1)/corr_colsum 
hist(normweightGOClust1)


# Use function, wcorrCluster, to get the weighted values for all blinded genes in cluster 1
cluster1wGO <- wcorrCluster(corrblind1, corrCluster1, cluster1GO)

# Determining threshold based on histogram
hist(cluster1wGO$YPR002W)
inpGO <- names(which(normweightGOClust1 > 0.1))

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster1wGOdf <- as.data.frame(do.call(rbind, cluster1wGO))
cluster1wGOdf <- cluster1wGOdf[order(rownames(cluster1wGOdf)),]
cluster1wGOdf <- cluster1wGOdf[,order(colnames(cluster1wGOdf))]

# using 0.1 as threshold
Clust1thresh <- 0.005

inpClust1df2 <- (as.matrix(cluster1wGOdf) > Clust1thresh)*1
inpClust1df <- (as.matrix(cluster1wGOdf) > Clust1thresh)*1



# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster1inp <- orgGOmatrixwide[orgGOmatrixwide$GeneID %in% rownames(inpClust1df),]
rownames(cluster1inp) <- cluster1inp[,1]
cluster1inp[,1] = c()
# Filter by column
cluster1inp <- cluster1inp[,colnames(cluster1inp) %in% colnames(inpClust1df)]
# Order by row names and column names
cluster1inp1 <- cluster1inp[order(rownames(cluster1inp)),]
cluster1inp1 <- cluster1inp[,order(colnames(cluster1inp))]


# Filter inpClust1df object to include the genes and GO terms of cluster1inp object above
inpClust1trim <- inpClust1df[rownames(inpClust1df2) %in% rownames(cluster1inp),]
# Order by row names for direct comparison
inpClust1trim <- inpClust1trim[order(rownames(inpClust1trim)),]


# Comparison between input df and resultant df
Clust1inpVSinp <- ifelse(cluster1inp == inpClust1trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.005
Clust1neg <- inpClust1trim - cluster1inp
which(Clust1neg == -1)

```
</br>

### Using weightcorr2Clust
```{r echo=TRUE}

cluster1GO2 <- weightcorr2Clust(corrblind1, corrCluster1, cluster1GO)

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster1GO2df <- as.data.frame(do.call(rbind, cluster1GO2))
cluster1GO2df <- cluster1GO2df[order(rownames(cluster1GO2df)),]
cluster1GO2df <- cluster1GO2df[,order(colnames(cluster1GO2df))]

# using 0.01 as threshold
Clust1thresh <- 0.01

inpClust1df <- as.data.frame(sapply(cluster1GO2df, function(x){
   as.numeric(x > Clust1_threshold)
  }))
row.names(inpClust1df) <- row.names(cluster1GO2df)

# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster1inp <- orgGOmatrixwide[orgGOmatrixwide$GeneID %in% rownames(inpClust1df),]
rownames(cluster1inp) <- cluster1inp[,1]
cluster1inp[,1] = c()
# Filter by column
cluster1inp <- cluster1inp[,colnames(cluster1inp) %in% colnames(inpClust1df)]
# Order by row names and column names
cluster1inp1 <- cluster1inp[order(rownames(cluster1inp)),]
cluster1inp1 <- cluster1inp[,order(colnames(cluster1inp))]


# Filter inpClust1df object to include the genes and GO terms of cluster1inp object above
inpClust1trim <- inpClust1df[rownames(inpClust1df) %in% rownames(cluster1inp),]
# Order by row names for direct comparison
inpClust1trim <- inpClust1trim[order(rownames(inpClust1trim)),]
inpClust1trim <- inpClust1trim[,order(colnames(inpClust1trim))]

# Comparison between input df and resultant df
Clust1inpVSinp2 <- ifelse(cluster1inp == inpClust1trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.01
Clust1neg <- inpClust1trim - cluster1inp
which(Clust1neg == -1)

```
</br>

### Using cluster 3
```{r echo=TRUE}

corrCluster3 <- corrAllClusters$Cluster3
clust3GO <- GOallClusters$Cluster3
cluster3_list <- rownames(clust3GO)
# Make a function to list all blind genes in a cluster
clust3blind <- cluster3_list[cluster3_list %in% dictBlind$blind]

cluster3wGO2 <- weightcorr2Clust(clust3blind, corrCluster3, clust3GO)

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster3wGO2df <- as.data.frame(do.call(rbind, cluster3wGO2))
cluster3wGO2df <- cluster3wGO2df[order(rownames(cluster3wGO2df)),]
cluster3wGO2df <- cluster3wGO2df[,order(colnames(cluster3wGO2df))]

# using 0.01 as threshold
Clust3thresh <- 0.01

impClust3df <- as.data.frame(sapply(cluster3wGO2df, function(x){
   as.numeric(x > Clust3thresh)
  }))
impClust3df <- t(impClust3df)
row.names(impClust3df) <- row.names(cluster3wGO2df)

# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster3inp <- orgGOmatrixwide[orgGOmatrixwide$GeneID %in% rownames(impClust3df),]
rownames(cluster3inp) <- cluster3inp[,1]
cluster3inp[,1] = c()
# Filter by column
cluster3inp <- cluster3inp[,colnames(cluster3inp) %in% colnames(impClust3df)]
# Order by row names and column names
cluster3inp <- cluster3inp[order(rownames(cluster3inp)),]
cluster3inp <- cluster3inp[,order(colnames(cluster3inp))]


# Filter impClust3df object to include the genes and GO terms of cluster1inp object above
inpClust3trim <- impClust3df[rownames(impClust3df) %in% rownames(cluster3inp),]

# Comparison between input df and resultant df
Clust3inpVSinp3 <- ifelse(cluster3inp == inpClust3trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.01
Clust3neg <- inpClust3trim - cluster3inp
which(Clust3neg == -1)
```

</br>


### 13. Gene Enrichment Analysis 
```{r echo=TRUE}
# ---------Functional Enrichment Analysis using using goProfileR

# Gene enrichment query
goResults_1 <- gprofiler(query = genesOfInterest,
                     organism = 'org',
                     hier_filtering = 'none')

## Subset and reorder gProfiler results to only include columns of interest
gprofiler_results_oe_reordered <- goResults_1[, c("term.id", "domain", "term.name", "p.value", "overlap.size", "term.size", "intersection")]

# Order the results by p-adjusted value
gprofiler_results_oe_reordered <- gprofiler_results_oe_reordered[order(gprofiler_results_oe_reordered$p.value), ]
 
# Extract only the 'GO' terms from the results
gprofiler_results_oe_GOs <- gprofiler_results_oe_reordered[grep('GO:', gprofiler_results_oe_reordered$term.id), ]

# Write the enriched GO results to file
write.csv(gprofiler_results_oe_GOs,
            "Data/gprofiler_cluster1.csv")


# --------Over-representation analysis of biological processes
# perform Gene Set Enrichment Analysis (GSEA) using gage package

columns(org.Sc.sgd.db)
keys(org.Sc.sgd.db)
keytypes(org.Sc.sgd.db)

# The input to gage must be Entrez identifiers so we first need to map our gene symbols to Entrez
ann <- mapIds(org.Sc.sgd.db, keys=rownames(cluster1), column="ENTREZID", keytype="ENSEMBL")
cluster1_non_na <- names(!is.na(ann))
fc <- normAgg[rownames(normAgg) %in% cluster1_non_na,]

table(is.na(fc))

# Create a mapping between gene identifiers and GO terms.
allGo <- as.list(GOTERM)
bp <- Filter(function(go) go@Ontology == 'BP', allGo)
goMapping <- mget(names(bp), org.Sc.sgdGO2ALLORFS, ifnotfound=NA)

# gene set enrichment analysis
goFc <- gage(fc, goMapping)
goGreater <- as.data.frame(goFc$greater)
goGreater <- goGreater[goGreater$q.val < 0.01 & !is.na(goGreater$q.val),]
goTermNames <- lapply(mget(rownames(goGreater), GOTERM), function(go) go@Term)
goGreater$Term <- goTermNames[rownames(goGreater)]
goGreater



#---------using clusterprofiler
library("clusterProfiler")


background_genes <- clusters[clusters$ClusterNumber !=1,]
background_geneIDs <- rownames(agg)

orgCPGO <- enrichGO(genesOfInterest, "org.Sc.sgd.db", keyType = "ENSEMBL", ont = "ALL", universe = background_geneIDs)
orgCPGOdf <- as.data.frame(orgCPGO)
orgCPGOdf
```

</br>


### 14. Assessing clustering tendency

```{r echo=TRUE}

# Hopkins statistic: If the value of Hopkins statistic is close to 1 (far above 0.5), 
# then we can conclude that the dataset is significantly clusterable

# Visual approach: The visual approach detects the clustering tendency by counting 
# the number of square shaped dark (or colored) blocks along the diagonal in 
# the ordered dissimilarity image.

gradient.color <- list(low = "steelblue",  high = "white")

# create a 5 x 5 inch png image
png("Data/clustering_tendency_PASS.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)          
normAgg %>% get_clust_tendency(n = nrow(normAgg)-1, gradient = gradient.color)

dev.off()
```
</br>

### 15. K-means Clustering

```{r echo=TRUE}
# determining the number of clusters (K) using NbClust package

#val_K_clust <- NbClust(normAgg, min.nc=1, max.nc=100, method="kmeans")
num_K_clust <- fviz_nbclust(normAgg, FUNcluster = kmeans, method = "wss", print.summary = TRUE)

# K-means Clustering

# As k-means clustering algorithm starts with k randomly selected centroids, it’s always recommended to use the set.seed() function in order to set a seed for R’s random number generator

set.seed(123)
km.res_100 <- kmeans(normAgg, 100, nstart = 25, iter.max = 50)

png("Data/K_means_PASS.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)
fviz_cluster(km.res_100, data = normAgg, geom="point")
dev.off()
```
</br>

### 16. Applying Function for K-means results
```{r echo=TRUE}

```


</br>

### Session Information 

```{r Session Info, echo=FALSE}
sessionInfo()
```